jf_agent/git/gitlab_adapter.py

```diff
@@ -0,0 +1,453 @@
+                'No projects found.  Make sure your token has appropriate access to GitLab.'
+            )
+        return projects
+
+    @diagnostics.capture_timing()
+    @agent_logging.log_entry_exit(logger)
+    def get_users(self, include_project_ids: List[int]) -> List[NormalizedUser]:
+        print('downloading gitlab users... ', end='', flush=True)
+        users = [
+            _normalize_user(user)
+            for project_id in include_project_ids
+            for user in self.client.list_group_members(project_id)
+        ]
+        print('✓')
+        return users
+
+    @agent_logging.log_entry_exit(logger)
+    def get_repos(
+        self,
+        normalized_projects: List[NormalizedProject],
+        include_repos_ids: List[int],
+        exclude_repos_ids: List[int],
+        redact_names_and_urls: bool,
+    ) -> List[NormalizedRepository]:
+        print('downloading gitlab repos... ', end='', flush=True)
+
+        nrm_repos: List[NormalizedRepository] = []
+        for nrm_project in normalized_projects:
+
+            for i, api_repo in enumerate(
+                tqdm(
+                    self.client.list_group_projects(nrm_project.id),
+                    desc=f'downloading repos for {nrm_project.name}',
+                    unit='commits',
+                ),
+                start=1,
+            ):
+                if len(include_repos_ids) > 0 and api_repo.id not in include_repos_ids:
+                    continue  # skip this repo
+                if len(exclude_repos_ids) > 0 and api_repo.id in exclude_repos_ids:
+                    continue  # skip this repo
+
+                nrm_branches = self.get_branches(api_repo, redact_names_and_urls)
+                nrm_repos.append(
+                    _normalize_repo(api_repo, nrm_branches, nrm_project, redact_names_and_urls)
+                )
+
+        print('✓')
+        if not nrm_repos:
+            raise ValueError(
+                'No repos found. Make sure your token has appropriate access to GitLab and check your configuration of repos to pull.'
+            )
+        return nrm_repos
+
+    @diagnostics.capture_timing()
+    @agent_logging.log_entry_exit(logger)
+    def get_branches(self, api_repo, redact_names_and_urls) -> List[NormalizedBranch]:
+        try:
+            return [
+                _normalize_branch(api_branch, redact_names_and_urls)
+                for api_branch in self.client.list_project_branches(api_repo.id)
+            ]
+        except requests.exceptions.RetryError as e:
+            repo_name = (
+                api_repo.name
+                if not redact_names_and_urls
+                else _repo_redactor.redact_name(api_repo.name)
+            )
+            logging.warning(
+                f"Received a RetryError {e} when pulling branches from '{repo_name}'"
+                "This is most likely because no repo setup to pull branches from GitlabProject -- will treat like there are no branches"
```

Comment:
Doesn't read quite right

Reply:
This should probably be sent to stdout via `print` rather than to the logger, since it could contain customer-sensitive info (repo names), as per the comments in `agent_logging.py`.

This same comment probably applies to most or all of the other warnings being written in the GitLab adapter; probably best to just standardize on them all going to stdout only.

Reply:
Ah fair point. Although the repo name is redacted prior to being logged. I will change this over to `print()`